core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java

```diff
@@ -0,0 +1,297 @@
+        if (fetchResult.hasData() == false) {
+            return getFailedRerouteExplanation(explain, allocation, "shard state data for " + shardId + " is still being fetched");
+        }
+
+        if (node != null) {
+            TransportNodesListGatewayStartedShards.NodeGatewayStartedShards nodeShardState = fetchResult.getData().get(discoNode);
+            if (nodeShardState == null) {
+                return getFailedRerouteExplanation(explain, allocation, "found no node with shard data for shard " + shardId);
+            }
+
+            if (nodeShardState.storeException() != null || (nodeShardState.version() == -1 && nodeShardState.allocationId() == null)) {
+                return getFailedRerouteExplanation(explain, allocation, "found no shard data for shard " + shardId + " on node " + node);
+            }
+
+            if (allocationId != null) {
+                if (allocationId.equals(nodeShardState.allocationId()) == false) {
+                    return getFailedRerouteExplanation(explain, allocation,
+                        "found shard data for shard " + shardId + " on node " + node + " but does not match allocation id (expected ["
+                            + allocation + "] but was [" + nodeShardState.allocationId() + "])");
+                }
+            }
+        } else {
+            if (allocationId != null) {
+                // search for shard with allocation id on all nodes
+                for (Map.Entry<DiscoveryNode, TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> entry : fetchResult.getData().entrySet()) {
+                    if (entry.getValue().storeException() == null && allocationId.equals(entry.getValue().allocationId())) {
+                        discoNode = entry.getKey();
+                        break;
+                    }
+                }
+                if (discoNode == null) {
+                    return getFailedRerouteExplanation(explain, allocation,
+                        "found no shard data for shard " + shardId + " matching allocation id [" + allocationId + "]");
+                }
+            }
+        }
+
+        assert discoNode != null;
+
+        // OK, now we go ahead and allocate
+        RoutingNode routingNode = routingNodes.node(discoNode.id());
+        if (routingNode == null) {
+            if (!discoNode.dataNode()) {
+                return getFailedRerouteExplanation(explain, allocation,
+                    "Allocation can only be done on data nodes, not [" + node + "]");
+            } else {
+                return getFailedRerouteExplanation(explain, allocation,
+                    "Could not find [" + node + "] among the routing nodes");
+            }
+        }
+
+        Decision decision = allocation.deciders().canAllocate(shardRouting, routingNode, allocation);
+        if (ignoreDeciders == false && decision.type() == Decision.Type.NO) {
```

Comment:
The idea behind that was so that we could see in the RerouteExplanation if it was Yes, Forced Yes or Throttle. It's probably simpler to just always provide the decision  (YES, "override deciders") in case of explicit override.